### PR TITLE
Add symnode magic method testing

### PR DESCRIPTION
**Pull request by albanD**
There are failures that need to be addressed before landing:
- Some issue with handling of booleans.
- Most functions return wrong result when mixing int/float

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -271,6 +271,11 @@ def _nyi():
 
 float_magic_methods = {"add", "sub", "mul", "truediv", "ceil", "floor", "eq", "gt", "lt", "le", "ge", "pow"}
 
+magic_methods_on_builtins = {"min", "max"}
+magic_methods_on_math = {"min", "max"}
+always_float_magic_methods = {"truediv", "sym_float"}
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:276]: Current truediv leads to
```python
======================================================================
ERROR: test_method_fn_floordiv_first_type_int_second_type_float (__main__.TestSymNumberMagicMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/albandes/local/pytorch/3.8_debug_source/torch/testing/_internal/common_utils.py", line 248, in instantiated_test
    test(self, **param_kwargs)
  File "test_dynamic_shapes.py", line 538, in test_method
    self._do_test(fn, inp1, inp2, shape_env, is_unary_fn)
  File "test_dynamic_shapes.py", line 498, in _do_test
    self.assertEqual(guard_fn(out), ref_out)
  File "test_dynamic_shapes.py", line 479, in guard_fn
    raise e
  File "test_dynamic_shapes.py", line 474, in guard_fn
    return getattr(v.node, f"guard_{tp}")("", 0)
  File "/home/albandes/local/pytorch/3.8_debug_source/torch/fx/experimental/symbolic_shapes.py", line 188, in guard_int
    return int(self.shape_env.evaluate_expr(self.expr))
  File "/home/albandes/local/pytorch/3.8_debug_source_env/lib/python3.8/site-packages/sympy/core/expr.py", line 321, in __int__
    r = self.round(2)
  File "/home/albandes/local/pytorch/3.8_debug_source_env/lib/python3.8/site-packages/sympy/core/expr.py", line 3817, in round
    raise TypeError(
TypeError: Expected a number but got FloorDiv:
```

Also current impl doesn't seem to follow best practices as it does fusion which is explicitly mentioned as a bad idea: https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -242,7 +202,7 @@ def eval(cls, base, divisor):
     'mod': lambda a, b: a % b,
     'pow': lambda a, b: a ** b,
     'truediv': lambda a, b: a / b,
-    'floordiv': lambda a, b: FloorDiv(a, b),
+    'floordiv': lambda a, b: a // b,
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:205]: @Chillee 
Current floordiv leads to pretty nasty error. So I am proposing to remove it. WDYT
```python
======================================================================
ERROR: test_method_fn_floordiv_first_type_int_second_type_float (__main__.TestSymNumberMagicMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/albandes/local/pytorch/3.8_debug_source/torch/testing/_internal/common_utils.py", line 248, in instantiated_test
    test(self, **param_kwargs)
  File "test_dynamic_shapes.py", line 538, in test_method
    self._do_test(fn, inp1, inp2, shape_env, is_unary_fn)
  File "test_dynamic_shapes.py", line 498, in _do_test
    self.assertEqual(guard_fn(out), ref_out)
  File "test_dynamic_shapes.py", line 479, in guard_fn
    raise e
  File "test_dynamic_shapes.py", line 474, in guard_fn
    return getattr(v.node, f"guard_{tp}")("", 0)
  File "/home/albandes/local/pytorch/3.8_debug_source/torch/fx/experimental/symbolic_shapes.py", line 188, in guard_int
    return int(self.shape_env.evaluate_expr(self.expr))
  File "/home/albandes/local/pytorch/3.8_debug_source_env/lib/python3.8/site-packages/sympy/core/expr.py", line 321, in __int__
    r = self.round(2)
  File "/home/albandes/local/pytorch/3.8_debug_source_env/lib/python3.8/site-packages/sympy/core/expr.py", line 3817, in round
    raise TypeError(
TypeError: Expected a number but got FloorDiv:
```

Also current impl doesn't seem to follow best practices as it does fusion which is explicitly mentioned as a bad idea: https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval

Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:205]: 🤔 This error looks like somehow, `evaluate_expr` is returning a sympy expression and not an integer.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:205]: This evaluate expr business, where we ask sympy to evaluate int/float ops for us, is awfully questionable. We should just compute it ourselves

Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:205]: I don't see why it should be questionable 🤔

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:205]: It's just a complicated way of doing something we can do very simply

Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:205]: I don't see how we do it simpler? `evaluate_expr` (other than adding a guards) simply substitutes in all of the symbols.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:205]: If we only have backed symints, we can just maintain the true value for every intermediate compute. Even with unbacked symints we could maintain the unsimplified expression and eval it like a normal interpreter

**Comment by ezyang** [comment]
@Chillee on FloorDiv removal

**Comment by Chillee** [comment]
@ezyang I don't think we can remove `FloorDiv`. The two biggest issues I'm aware of are:

1. `sympy.floordiv` results in intermediate float computations, which are quite annoying and prevent simplifications.
2. Divide itself does some pretty weird simplifications.

image

**Review by ezyang — CHANGES_REQUESTED** [review]

File: test/test_dynamic_shapes.py [hunk 3]
```diff
@@ -386,6 +394,148 @@ def forward(self, a_1: f32[s0, s1], b_1: f32[s2, s1]):
...
+
+    print("expected_failure_sym_magic_methods = {")
+    print("\n".join(out))
+    print("}")
+
+if COLLECT_EXPECT:
+    atexit.register(print_seen)
+
+expected_failure_sym_magic_methods = {
+    ('floordiv', 'int', 'SymFloat'),  # unsupported operand type(s) for //: 'int' and 'SymFloat'
```
[the remark this answers is not in the document]
Comment (albanD) [reply test/test_dynamic_shapes.py:419]: This is allowed in python:
```
>>> 2 // 1.5
1.0
```

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -156,7 +156,7 @@ def wrap_int(self, num):
 
     def wrap_float(self, num):
         assert isinstance(num, float)
-        return SymNode(sympy.Integer(num), self.shape_env, float, constant=num)
+        return SymNode(sympy.Float(num), self.shape_env, float, constant=num)
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:159]: This is the ONLY logic change left in this PR. All other issues are skipped in the test

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -271,6 +271,14 @@ def _nyi():
 
 float_magic_methods = {"add", "sub", "mul", "truediv", "ceil", "floor", "eq", "gt", "lt", "le", "ge", "pow"}
 
+magic_methods_on_builtins = {"min", "max"}
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:274]: These are just moved around to be re-used in the testing code. No logic change.

**Review by ezyang — APPROVED** [review]
only skimmed over the testing code

**Pull request closed without merging**